### PR TITLE
Fix schedule assember bug

### LINKED ISSUE #3899: Pulse assember overwrite pulse from another schedule
### Information

- **Qiskit Terra version**:
- **Python version**:
- **Operating system**:

### What is the current behavior?

When we have two different pulses with the same name in different schedules, assembler combines those pulses.

### Steps to reproduce the problem

```
from qiskit.test.mock.backends import FakeAlmaden
import qiskit

backend = FakeAlmaden()

schedules = []
for amp in (0.1, 0.2):
    sched = qiskit.pulse.Schedule()
    sched += qiskit.pulse.pulse_lib.gaussian(duration=100, amp=amp, sigma=30, name='my_gaussian')(qiskit.pulse.DriveChannel(0))
    sched += qiskit.scheduler.utils.measure(qubits=(0, ), backend=backend) << 100
    schedules.append(sched)

qobj = qiskit.assemble(schedules, backend)

for pulse_item in qobj.config.pulse_library:
    print(pulse_item.name)
```
this outputs
```
my_gaussian
M_m0_5637
```
you can see two `my_gaussian` pulses with different samples are combined.

### What is the expected behavior?

The second `my_gaussian` pulse should be renamed and independently registered to the pulse library.

### PR DESCRIPTION
**Pull request by nkanazawa1989**
### Summary
fix #3899 

### Details and comments
new unittest is added.

### CODE REVIEW

File: qiskit/assembler/assemble_schedules.py [hunk 1]
```diff
@@ -154,14 +157,15 @@ def _assemble_instructions(
...
+        user_pulselib: User pulse library from previous schedule.
 
     Returns:
         A list of converted instructions, the user pulse library dictionary (from pulse name to
         pulse command), and the maximum number of readout memory slots used by this Schedule.
     """
     max_memory_slot = 0
     qobj_instructions = []
-    user_pulselib = {}
+    user_pulselib = deepcopy(user_pulselib)
```
Comment (lcapelluto) [inline qiskit/assembler/assemble_schedules.py]: Can this be avoided? Can get slow if there are many pulses. Would it work if this line was removed?

**Comment by lcapelluto** [comment]
Thank you for catching this! 🐛

File: qiskit/assembler/assemble_schedules.py (hunk 1 above)
Comment (lcapelluto) [reply qiskit/assembler/assemble_schedules.py]: I checked out locally, this line and line 108 can be removed, and `user_pulses` -> `user_pulse_lib` in line 103

Comment (nkanazawa1989) [reply qiskit/assembler/assemble_schedules.py]: ok. I thought you don't prefer such mutability. I'll remove this.

File: qiskit/assembler/assemble_schedules.py [hunk 2]
```diff
@@ -144,8 +144,9 @@ def _assemble_experiments(
 def _assemble_instructions(
         schedule: Schedule,
         instruction_converter: InstructionToQobjConverter,
-        run_config: RunConfig
-) -> Tuple[List[PulseQobjInstruction], Dict[str, Command], int]:
+        run_config: RunConfig,
+        user_pulselib: Dict[str, Command]
+) -> Tuple[List[PulseQobjInstruction], int]:
     """Assembles the instructions in a schedule into a list of PulseQobjInstructions and returns
     related metadata that will be assembled into the Qobj configuration.
```
Comment (lcapelluto) [inline qiskit/assembler/assemble_schedules.py]: Can you update the documentation to say that the input is modified?

Comment (nkanazawa1989) [reply qiskit/assembler/assemble_schedules.py]: Is this what you intended?
https://github.com/Qiskit/qiskit-terra/pull/3900/commits/477f6fb68ed958b66754cd6a761a3646ff912615

**Comment by taalexander** [comment]
Any update on this @lcapelluto?

**Pull request merged**